src/iarc7_vision/roomba_blob_vision_node.py

```diff
@@ -0,0 +1,351 @@
+
+        # Starup the loop
+        try:
+            rospy.spin()
+        except KeyboardInterrupt:
+            rospy.loginfo("Shutting down")
+        
+    def callback(self, data):
+        """
+        Receive the callback from the camera, process the image, and publish
+        the roomba locations found
+
+        :param data: Image
+        :type data: sensor_msgs.msg.Image
+        :return: None
+        """
+        try:
+            cv_image = self.bridge.imgmsg_to_cv2(data, "bgr8")
+            # cv_image = cv2.resize(cv_image,None,fx=0.5,fy=0.5, 
+            #                       interpolation=cv2.INTER_CUBIC)
+            trans = self.tf_buffer.lookup_transform('map',
+                                'bottom_camera_optical', rospy.Time(0))
+            self.filter_frame(cv_image, trans)
+        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, \
+                tf2_ros.ExtrapolationException, CvBridgeError) as e:
+            rospy.logerr(e)
+
+
+    def filter_frame(self, frame, trans):
+        """
+        Applies a four-step algorithm to find all the roombas in a given BGR image.
+    
+        :param frame: raw BGR image in which to find roombas
+        :type frame: numpy.ndarray of shape 3 x WIDTH x HEIGHT with dtype uint8
+        :return: None
+    
+        .. note::
+            Does not return data, only debugs the result
+    
+        .. note::
+            The image origin is in the upper-left hand corner
+        """
+        points = []
+    
+        for img_coords in self.bound_roombas(frame):
+            cam_ray = Vector3Stamped()
+            cam_ray.vector.x, cam_ray.vector.y, cam_ray.vector.z = self.camera.projectPixelTo3dRay(img_coords)
+            
+            # Convert that camera ray to world space (map frame)
+            map_ray = tf2_geometry_msgs.do_transform_vector3(cam_ray, copy.deepcopy(trans)) # reassigning
+            
+            # Scale the direction to hit the ground (plane z=0)
+            direction_scale = (trans.transform.translation.z - ROOMBA_HEIGHT) / map_ray.vector.z
+            roomba_pos = Point()
+            roomba_pos.x = trans.transform.translation.x - map_ray.vector.x * direction_scale
+            roomba_pos.y = trans.transform.translation.y - map_ray.vector.y * direction_scale
+            roomba_pos.z = 0
+    
+            # Debug the roomba line
+            points.append(trans.transform.translation)
+            points.append(roomba_pos)
+
+            # Add the roomba to array and publish
+            # TODO This can be improved
+            sq_tolerance = 0.1 if len(self.roombas) < 10 else 1000
+            index = -1
+            for i in xrange(len(self.roombas)):
+                if (roomba_pos.x - self.roombas[i].x)**2 + (roomba_pos.y - self.roombas[i].y)**2 < sq_tolerance:
+                    self.roombas[i] = roomba_pos
+                    index = i
+                    break
+            if index == -1:
```

Comment:
Just a tip, you could use Python's `for...else` construct here, it's designed for exactly this situation